Libraries/VLM/Models/Paligemma.swift

```diff
@@ -0,0 +1,727 @@
+        let embedDimension = imageFeatures.dim(2)
+        let (batchSize, sequenceLength) = inputIds.shape2
+        var scaledImageFeatures = imageFeatures / pow(Float(config.hiddenSize), 0.5)
+        var finalEmbedding = zeros([batchSize, sequenceLength, embedDimension])
+
+        let textMask = (inputIds .!= config.imageTokenIndex) & (inputIds .!= config.padTokenId)
+        let imageMask = inputIds .== config.imageTokenIndex
+        let padMask = inputIds .== config.padTokenId
+
+        // expand masks to match embedding dimension
+        var textMaskExpanded = expandedDimensions(textMask, axis: -1)
+        textMaskExpanded = repeated(textMaskExpanded, count: embedDimension, axis: -1)
+        var padMaskExpanded = expandedDimensions(padMask, axis: -1)
+        padMaskExpanded = repeated(padMaskExpanded, count: embedDimension, axis: -1)
+
+        // insert padding and text token embeddings
+        finalEmbedding = which(textMaskExpanded, inputEmbedding, finalEmbedding)
+        finalEmbedding = which(padMaskExpanded, zeros(like: finalEmbedding), finalEmbedding)
+
+        let padSize = finalEmbedding.dim(1) - scaledImageFeatures.dim(1)
+        scaledImageFeatures = padded(scaledImageFeatures, widths: [0, .init((0, padSize)), 0])
+
+        // insert image embeddings - the image mask is always less or equal to the sentence in length
+        var imageMaskExpanded = expandedDimensions(imageMask, axis: -1)
+        imageMaskExpanded = repeated(imageMaskExpanded, count: embedDimension, axis: -1)
+        finalEmbedding = which(imageMaskExpanded, scaledImageFeatures, finalEmbedding)
+
+        finalEmbedding = which(padMaskExpanded, zeros(like: finalEmbedding), finalEmbedding)
+
+        let attentionMaskExpanded1 = expandedDimensions(attentionMask, axis: 1)
+        let attentionMaskExpanded2 = expandedDimensions(attentionMask, axis: 2)
+        var finalAttentionMask4d = attentionMaskExpanded1 * attentionMaskExpanded2
+        finalAttentionMask4d = expandedDimensions(finalAttentionMask4d, axis: 1)
+
+        return (finalEmbedding, finalAttentionMask4d)
+    }
+
+    public func prepare(_ input: LMInput, cache: [any KVCache], windowSize: Int?) throws
+        -> PrepareResult
+    {
+        guard let image = input.image else { throw VLMError.imageRequired }
+        guard let mask = input.text.mask else { throw VLMError.maskRequired }
+        let inputIds = input.text.tokens
+
+        let (inputEmbedding, finalAttentionMask4d) = inputEmbeddings(
+            inputIds: inputIds, pixelValues: image.pixels, mask: mask)
+
+        let result = languageModel(
+            inputIds, cache: cache, inputEmbedding: inputEmbedding, mask: finalAttentionMask4d)
+
+        return .logits(result)
+    }
+
+    public func callAsFunction(_ inputs: MLXArray, cache: [any KVCache]?) -> MLXArray {
+        languageModel(inputs, cache: cache).logits
+    }
+}
+
+private func loadConfiguration(url: URL) throws -> PaliGemma {
```

Comment:
These next couple of functions are just stubs to let me try it out -- this will work much like the LLM models